### PR TITLE
Added db.aggregate method

### PR DESCRIPTION
**Pull request by rozza**
JAVA-3116

Took the same approach as the watch method and introduced a new enum to describe the level at which the operation takes place.

This includes the updated json that was approved in SPEC-1184.

Patch Build: https://evergreen.mongodb.com/version/5c334fbb2a60ed24ead2f23a

### CODE REVIEW

**Comment by rozza** [comment]
Patch: https://evergreen.mongodb.com/version/5c1bad0f2fbabefdfdaa03b0

File: driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java [hunk 1]
```diff
@@ -108,6 +111,19 @@ BsonDocument getOperationResults(final BsonDocument operation, @Nullable final C
         }
     }
 
+    private String createMethodName(final String name, final String object) {
+        StringBuilder builder = new StringBuilder();
+        builder.append("get");
+        if (!object.isEmpty() && !object.equals("collection")) {
```
Comment (rozza) [inline driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java]: We only inject the object name if its set in the json file and not the default - so we now have `getAggregateResult` and `getDatabaseAggregateResult`

File: driver-core/src/test/resources/crud/db/db-aggregate.json [hunk 2]
```diff
@@ -0,0 +1,149 @@
+{
+  "database_name": "admin",
+  "data": [],
+  "minServerVersion": "3.6",
```
[the remark this answers is not in the document]
Comment (rozza) [reply driver-core/src/test/resources/crud/db/db-aggregate.json]: https://jira.mongodb.org/browse/SPEC-1198

File: driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java [hunk 3]
```diff
@@ -108,6 +111,19 @@ BsonDocument getOperationResults(final BsonDocument operation, @Nullable final C
         }
     }
 
+    private String createMethodName(final String name, final String object) {
```
Comment (rozza) [inline driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java]: We only inject the object name if its set in the json file and not the default (collection) - so we now have `getAggregateResult` and `getDatabaseAggregateResult`

File: driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java [hunk 4]
```diff
@@ -247,6 +263,44 @@ BsonDocument getAggregateResult(final BsonDocument collectionOptions, final Bson
         return toResult(iterable);
     }
 
+    BsonDocument getDatabaseAggregateResult(final BsonDocument collectionOptions, final BsonDocument arguments,
+                                            @Nullable final ClientSession clientSession) {
+        assumeTrue(!isSharded());
```
Comment (rozza) [inline driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java]: These tests don't work on sharded clusters

File: driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java [hunk 5]
```diff
@@ -247,6 +263,44 @@ BsonDocument getAggregateResult(final BsonDocument collectionOptions, final Bson
...
+        }
+        if (arguments.containsKey("batchSize")) {
+            iterable.batchSize(arguments.getNumber("batchSize").intValue());
+        }
+        if (arguments.containsKey("collation")) {
+            iterable.collation(getCollation(arguments.getDocument("collation")));
+        }
+
+        BsonDocument results = toResult(iterable);
+        for (BsonValue result : results.getArray("result", new BsonArray())) {
```
Comment (rozza) [inline driver-async/src/test/functional/com/mongodb/async/client/JsonPoweredCrudTestHelper.java]: On direct connections the driver adds readPreference primaryPreferred to commands that don't have a set readPref. This also removes other metadata ($clusterTime, signature & keyId) so that from the results of the aggregation can be comparable.

File: driver-async/src/test/unit/com/mongodb/async/client/MongoDatabaseSpecification.groovy [hunk 6]
```diff
@@ -349,6 +351,111 @@ class MongoDatabaseSpecification extends Specification {
         thrown(IllegalArgumentException)
     }
 
+    def 'should create ChangeStreamIterable correctly'() {
```
Comment (rozza) [inline driver-async/src/test/unit/com/mongodb/async/client/MongoDatabaseSpecification.groovy]: These changestream tests were missing from the async database suite - so ported them across. Can split into a separate commit.

File: driver-async/src/test/unit/com/mongodb/async/client/MongoDatabaseSpecification.groovy [hunk 7]
```diff
@@ -349,6 +351,111 @@ class MongoDatabaseSpecification extends Specification {
...
+        thrown(IllegalArgumentException)
+
+        when:
+        database.watch([null]).into([]) { result, t -> }
+
+        then:
+        thrown(IllegalArgumentException)
+    }
+
+    def 'should create AggregateIterable correctly'() {
```
Comment (rozza) [inline driver-async/src/test/unit/com/mongodb/async/client/MongoDatabaseSpecification.groovy]: Async AggregateIterable tests

File: driver-sync/src/test/functional/com/mongodb/client/JsonPoweredCrudTestHelper.java [hunk 8]
```diff
@@ -751,4 +810,8 @@ WriteConcern getWriteConcern(final BsonDocument arguments) {
     ReadConcern getReadConcern(final BsonDocument arguments) {
         return new ReadConcern(ReadConcernLevel.fromString(arguments.getDocument("readConcern").getString("level").getValue()));
     }
+
+    boolean isSharded() {
+        return database.runCommand(Document.parse("{isMaster: 1}")).get("msg", "").equals("isdbgrid");
```
Comment (rozza) [inline driver-sync/src/test/functional/com/mongodb/client/JsonPoweredCrudTestHelper.java]: Can't use the helper from core as that requires a local mongod running on the default port. Which breaks the test when running against embedded as there is no local mongod on 127.0.0.1

File: driver-core/src/main/com/mongodb/client/model/AggregationLevel.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+package com.mongodb.client.model;
+
+/**
+ * The aggregation level
+ *
+ * Where an aggregation command should take place.
+ *
+ * @since 3.10
+ */
+public enum AggregationLevel {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/client/model/AggregationLevel.java]: For 3.x, this has to be public and in a non-internal package since AggregateOperation depends on it, but although it's a bit weird let's deprecate it since we won't want it public in 4.0.

Comment (rozza) [reply driver-core/src/main/com/mongodb/client/model/AggregationLevel.java]: Done and merged
